src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(null, value, null);
+        if (size == 0) {
+            head = node;
+            tail = node;
+        } else {
+            node.prev = tail;
+            tail = node;
+            node.prev.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkSizeAdd(index);
+        if (index == 0 && size == 0 || index > 0 && index == size) {
+            add(value);
+        } else {
+            link(value, index);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return find(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = find(index).value;
+        find(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = find(index).value;
+        unlink(find(index));
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> newNode = new Node<>(null, object, null);
+        for (int i = 0; i < size; i++) {
+            if (newNode.value == find(i).value
+                    || newNode.value != null && newNode.value.equals(find(i).value)) {
+                unlink(find(i));
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("Index out of bounds " + index);
+        }
+    }
+
+    private void checkSizeAdd(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Can't add element to index " + index);
+        }
+    }
+
+    private Node<T> find(int index) {
+        Node<T> node = head;
+        if (index == 0) {
+            return node;
+        }
+        if (index == size - 1) {
+            return tail;
+        }
+        if (index > (size >> 1)) {
+            node = tail;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private void link(T value, int index) {
+        Node<T> next = find(index);
+        Node<T> newNode = new Node<>(null, value, null);
```

Comment:
Again, you can immediately initialize the next and the first to do fewer actions later.